Reserve MPU regions

thapav opened the pull request:
This PR adds support to reserve MPU regions by the vendor. (CONFIG_MPU_RESERVED)
In such a scenario, the TizenRT MPU configurations will be done past the reserved MPU regions.
It adds the following API's to mpuinit.c
mpu_get_region_info(), mpu_set_region_info(), mpu_regs_initialize()

tizen-build commented:
Target : [971240f55bc8e5a5be913b7d2b02eca7f8863224] - Code Rule Check OK.

tizen-build commented:
Target : [971240f55bc8e5a5be913b7d2b02eca7f8863224] - Code Rule Check (C++) OK.

sunghan-chang commented on os/arch/Kconfig:
```diff
@@ -107,13 +107,27 @@ config ARCH_USE_MMU
 config ARCH_USE_MPU
 	bool "Enable MPU"
 	default n
+	select MPU_RESERVED
```
When A is mandatory for B, then we use "select" syntax. But I don't think this case is fit.
How about removing this select and adding "depends on ARCH_USE_MPU" in MPU_RESERVED?
Because MPU_RESERVED could be unused with ARCH_USE_MPU, but ARCH_USE_MPU must be enabled to use MPU_RESERVED.

thapav replied:
Yes, I agree that would be more apt. I will modify

sunghan-chang reviewed (commented):
My original request is how to add board-specific MPU region.
We list the MPU regions in os/include/tinyara/mpu.h. But it's the common header, not a board-specific header.
Adding board-specific MPU regions in the common header is not good. That's why I requested.

This pr only includes the way how to give the offset. No way how to add board-specific value.
Could you update me about this?

sunghan-chang commented:
In first commit description, teh -> the

sunghan-chang commented:
By verification? confirmation?, could you add the conditional to check the max region number for each architecture?
```
for ARMv8-M
if (MPU_REG_NUM_MAX >= CONFIG_ARMV8M_MPU_NREGIONS) {
  error
}
for ARMv7-M
...
```
I think we could add this in mkconfig.c to verify in compilation time.

jeongchanKim commented on os/arch/Kconfig:
```diff
@@ -107,13 +107,27 @@ config ARCH_USE_MMU
+	---help---
+		The number of MPU regions reserved by the vendor. Some boards use
+		MPU regions for board specific functionalities. In such a scenario,
+		the current MPU regions need to be updated so that vendor specfic
```
specfic -> specific

kishore-sn commented on os/include/tinyara/mpu.h:
```diff
@@ -47,9 +47,18 @@ enum {
 /* Enum for MPU region numbers */
 
 enum MPU_REG_NUM {
+	MPU_REG_NUM_RESERVE = CONFIG_MPU_RESERVED,
```
I think we can use CONFIG_MPU_RESERVED directly in the first place of the enum. 
There is no need to introduce a new variable called MPU_REG_NUM_RESERVE

kishore-sn commented on os/include/tinyara/mpu.h:
```diff
@@ -47,9 +47,18 @@ enum {
 #endif
 
+/* Number of reserved MPU regions for a board. This is defined by the vendor.
+ * If not, default value is taken to be 0.
```
This statement seems wrong. Default value of reserved is taken as -1, so that our own region starts from zero.

kishore-sn commented on os/arch/Kconfig:
```diff
@@ -107,13 +107,27 @@ config ARCH_USE_MMU
+if ARCH_USE_MPU
+
+config MPU_RESERVED
+	int "Number of reserved MPU regions"
```
Its better to use depends instead of "if ARCH_USE_MPU".

sunghan-chang commented on os/arch/arm/src/common/up_mpuinit.c:
```diff
@@ -50,11 +50,56 @@
+void mpu_regs_initialize(FAR struct mpu_regs *mpu)
+{
+
+#ifdef CONFIG_APP_BINARY_SEPARATION
```
We should use another way to get the value for each.
Currently 2nd block concerns the config of 1st block. This makes code complex.
In 2nd block, only the config of 2nd block should be concerned.
How about below?
```c
uint8_t offset;

offset = mpu->board...;

#ifdef config_common
#ifdef config_reload_time
offse += 3;
#else
offset += 1;
#endif
mpu->common.. = offset;
#endif

#ifdef config_app_separation
#ifdef config_reload_time
offse += 3;
#else
offset += 1;
#endif
mpu->app_regs = offset;
#endif
```

thapav replied:
Done

sunghan-chang commented on os/include/tinyara/mpu.h:
```diff
@@ -34,6 +34,16 @@
  ********************************************************************************/
 
+struct mpu_regs {
+	uint8_t board_mpu_regs;
```
What does the member mean? The number of mpu region for each? The highest number for each? The start number?

thapav replied:
this is the number of MPU reserved regions for the vendor

sunghan-chang commented:
Please add the code to check overflow mpu region like
```c
if (mpu->max >= CONFIG_ARMV7M_MPU_NUM)
    #error alkfjalfjal;f;alfj
```

thapav commented:
Updated the PR, Kindly check.

> Please add the code to check overflow mpu region like
> 
> ```c
> if (mpu->max >= CONFIG_ARMV7M_MPU_NUM)
>     #error alkfjalfjal;f;alfj
> ```

Since we cannot add compile time check using variables, I have added DEBUGASSERT for the same. 
But, we have a similar check in mpu_get_register_config_value() where the board asserts as well. I think this check is not needed. Let me know your opinion.

sunghan-chang commented on os/include/tinyara/mpu.h:
```diff
@@ -34,6 +34,18 @@
+	uint8_t app_regions;
+	uint8_t stackovf_regions;
+	uint8_t maxmpu_regions;
+};
```
I suggest below
```c
struct mpu_usages_s {
    uint8_t nregion_board_specific;
    uint8_t nregion_common_bin;
    uint8_t nregion_app_bin;
    uint8_t nregion_stackovf;
    uint8_t max_nregion;
};
```

sunghan-chang commented on os/arch/arm/src/common/up_mpuinit.c:
```diff
@@ -66,6 +106,20 @@
+	if (g_mpuregs.maxmpu_regions > CONFIG_ARMV8M_MPU_NREGIONS) {
+		DEBUGASSERT(g_mpuregs.maxmpu_regions < CONFIG_ARMV8M_MPU_NREGIONS);
+	}
+#endif
```
Let's reduce duplicated code.
```c
#ifdef CONFIG_V7
#define MPU_MAX_NREGION    CONFIG_...V7M...
#elif defined (CONFIG_V8)
#define 
#endif 

void up_mpuinitialize(void)
{
    ....

    if (g_mpu_... >= MPU_MAX_NREGION) {
        debugassert
    }
}
```
1. debugassert is enough? not assert? What happen when we try to set invalid mpu region?
2. In the conditional, don't we need to '='? I think it needs.

sunghan-chang commented on os/include/tinyara/mpu.h:
```diff
@@ -34,6 +34,18 @@
+	uint8_t maxmpu_regions;
+};
+
+extern struct mpu_regs g_mpuregs;
```
The name 'g_mpuregs' looks like "gloabl_mpu_registers". We already use `reg` in tcb_s for register. Let's rename it as `g_mpu_usages`? `g_mpu_region_info`? or something else..

sunghan-chang replied:
SAM recommends not to use global variable. Let's make a new function and call it from outside to get mpu region number information.
```
in mpu.h
enum mpu_region_usages {
   MPU_REGION_BOARD_SPECIFIC,
   ...
}

in up_mpuinit.c (common file)
static struct mpu_usages_s g_mpu_usages;

uint8_t mpu_get_region_info(enum mpu_region_usages usages)
{
     uint8_t nregion;

     switch (usages) {
     case MPU_REGION_BOARD_SPECIFIC:
          nregion = g_mpu_xxxx.nregion_board_specific;
           break;
      case ...
      }
   return nregion;
}
```

We need to make a set function for board specific value.

thapav replied:
modified

sunghan-chang commented on os/arch/arm/src/common/up_mpuinit.c:
```diff
@@ -50,11 +50,119 @@
+
+void mpu_regs_initialize(FAR struct mpu_usages_s *mpu)
+{
+
```
Could you remove this empty line?

sunghan-chang commented on os/arch/arm/src/common/up_mpuinit.c:
```diff
@@ -50,11 +50,119 @@
+ *
+ ****************************************************************************/
+
+void mpu_regs_initialize(FAR struct mpu_usages_s *mpu)
```
Let's remove `FAR`

sunghan-chang commented on os/arch/arm/src/common/up_mpuinit.c:
```diff
@@ -50,11 +50,119 @@
+		break;
+	case MPU_REGION_MAX:
+		g_mpu_region_usages.max_nregion = val;
+		break;
```
We set almost region_usages in this file except board specific. I think no need to support others.

sunghan-chang commented on os/include/tinyara/mpu.h:
```diff
@@ -34,61 +34,43 @@
 };
 
+/* Enum to to get MPU region number info for MPU usages */
+enum mpu_region_usages {
```
Could you add `_e` postfix to say the type?
`enum mpu_region_usages_e`

sunghan-chang commented on os/binfmt/binfmt_execmodule.c:
```diff
@@ -188,15 +188,17 @@ int exec_module(FAR struct binary_s *binp)
 #else
 	/* Complete RAM partition will be configured as RW region */
-	mpu_get_register_config_value(&rtcb->mpu_regs[0], MPU_REG_NUM_APP, (uintptr_t)binp->ramstart, binp->ramsize, false, true);
+	mpu_get_register_config_value(&rtcb->mpu_regs[0], nregion - 1, (uintptr_t)binp->ramstart, binp->ramsize, false, true);
```
Here is `nregion - 1` but in stack protection, you use `nregion`. Something wrong either here or stack protection.

thapav replied:
Stack protection needs to be updated

jeongchanKim commented on os/arch/arm/src/common/up_mpuinit.c:
```diff
@@ -50,11 +50,119 @@
+
+	uint8_t offset = mpu->nregion_board_specific;
+
+#ifdef CONFIG_SUPPORT_COMMON_BINARY
```
`CONFIG_SUPPORT_COMMON_BINARY` and `CONFIG_OPTIMIZE_APP_RELOAD_TIME` can be enabled only when `CONFIG_APP_BINARY_SEPARATION`.
So I think you can move line 87 in the above.

thapav replied:
But since we need to distinguish between the scenarios of configuring complete RAM partition vs that of (text, RO & data) separately, we need CONFIG_OPTIMIZE_APP_RELOAD_TIME
And to have a common binary usage , we need CONFIG_SUPPORT_COMMON_BINARY

jeongchanKim commented on os/arch/arm/src/common/up_mpuinit.c:
```diff
@@ -50,11 +50,119 @@
+
+#ifdef CONFIG_SUPPORT_COMMON_BINARY
+#ifdef CONFIG_OPTIMIZE_APP_RELOAD_TIME
+	offset += 3;
```
Could you add comment why you add 3 or 1 in case of each config?

thapav replied:
This has been supported while configuring MPU regions. So i think not needed here again.

sunghan-chang replied:
I found the `MPU_NUM_REGIONS` in <tinyara/mpu.h>. It looks the same purpose of this.
Could you check that below is right?
```c
#ifdef CONFIG_SUPPORT_COMMON_BINARY
     offset += MPU_NUM_REGIONS;
     mpu->nregion_common_bin = offset;
#endif 

#ifdef CONFIG_APP_BINARY_SEPARATION
     offset += MPU_NUM_REGIONS;
     mpu->nregion_app_bin = offset;
#endif 
```

jeongchanKim commented on os/arch/arm/src/common/up_mpuinit.c:
```diff
@@ -50,11 +50,119 @@
+void mpu_regs_initialize(FAR struct mpu_usages_s *mpu)
+{
+
+	uint8_t offset = mpu->nregion_board_specific;
```
`mpu` can be dereferencing.

jeongchanKim commented on os/arch/arm/src/common/up_mpuinit.c:
```diff
@@ -50,11 +50,119 @@
 /****************************************************************************
  * Pre-processor Definitions
  ****************************************************************************/
+static struct mpu_usages_s g_mpu_region_usages;
```
For `mpu_usages_s`, please include `tinyara/mpu.h`.

jeongchanKim commented on os/arch/arm/src/common/up_mpuinit.c:
```diff
@@ -50,11 +50,119 @@
+		break;
+	}
+
+	return nregion;
```
If there is not matched case, it returns trash value. Please add default or initialize the `nregion`.

kishore-sn commented on os/arch/arm/src/common/up_mpuinit.c:
```diff
@@ -66,6 +174,14 @@
+
+	/* Maximum MPU region number sanity check */
+	if (g_mpu_region_usages.max_nregion >= MPU_MAX_NREGION) {
+		DEBUGASSERT(MPU_MAX_NREGION < CONFIG_ARMV8M_MPU_NREGIONS);
```
Please add a check for ARMV7 also. Else it would cause compile error.

thapav commented:
Addressed the review comments

sunghan-chang commented on os/binfmt/binfmt_execmodule.c:
```diff
@@ -188,15 +188,17 @@ int exec_module(FAR struct binary_s *binp)
 	/* Initialize the MPU registers in tcb with suitable protection values */
 #ifdef CONFIG_ARM_MPU
 #ifdef CONFIG_OPTIMIZE_APP_RELOAD_TIME
+	uint8_t nregion = mpu_get_region_info(MPU_REGION_APP_BIN);
```
It looks we need to check the return value for 0.

sunghan-chang commented on os/board/rtl8721csm/src/component/soc/realtek/amebad/fwlib/ram_hp/rtl8721dhp_app_start.c:
```diff
@@ -1338,6 +1338,9 @@ extern void __libc_init_array(void);
 	app_mpu_nocache_init();
+
+	/* Initialize number of board mpu regions */
+	mpu_set_region_info(MPU_REGION_BOARD_SPECIFIC, 1);
```
Please split the commit for this change.

thapav replied:
done

sunghan-chang commented on os/include/tinyara/mpu.h:
```diff
@@ -101,4 +83,10 @@ bool up_mpu_check_active(uint32_t *mpu_regs);
 
 void up_mpu_disable_region(uint32_t *mpu_regs);
 
+void mpu_regs_initialize(FAR struct mpu_usages_s *mpu);
```
As I said before, regs could be "registers", not "regions" because of tcb_s. Let's write full word for region as you did for other function and variable.
```
void mpu_region_initialize....
```

sunghan-chang commented on os/arch/arm/src/common/up_mpuinit.c:
```diff
@@ -66,6 +162,14 @@
+
+	/* Maximum MPU region number sanity check */
+	if (g_mpu_region_usages.max_nregion >= MPU_MAX_NREGION) {
+		DEBUGASSERT(g_mpu_region_usages.max_nregion < MPU_MAX_NREGION);
```
if and debugassert are duplicated.
You can remove if conditional.

sunghan-chang commented on os/arch/arm/src/common/up_mpuinit.c:
```diff
@@ -50,11 +51,106 @@
+		g_mpu_region_usages.nregion_board_specific = val;
+		break;
+	}
+}
```
This function only supports board-specific. 
1. The usages argument looks redundant.
2. There is no way to get any result for invalid request.

How about changing the function to support only board-specific?
```
/* Set a number of mpu region for board-specific purpose */
void mpu_set_nregion_board_specific(uint8_t num)
{
    g_mpu_region_usages.nregion_board_specific = num;
}
```

sunghan-chang commented on os/arch/arm/src/common/up_mpuinit.c:
```diff
@@ -50,11 +51,106 @@
+	mpu->max_nregion = mpu->nregion_stackovf;
+#else
+	mpu->max_nregion = mpu->nregion_app_bin;
+#endif
```
How about below?
```
#ifdef CONFIG_STACK...
   offset += 1;
   mpu->.... = offset;
#endif 

   mpu->max_nregion = offset;
```

thapav commented:
Addressed the review comments, kindly check

sijupunnnoose commented on os/board/rtl8721csm/src/component/soc/realtek/amebad/fwlib/ram_hp/rtl8721dhp_app_start.c:
```diff
@@ -1338,6 +1338,9 @@ extern void __libc_init_array(void);
 	app_mpu_nocache_init();
+
+	/* Initialize number of mpu regions for board specific purpose */
+	mpu_set_nregion_board_specific(1);
```
How about having no of MPU regions as a config instead of hardcoding in code? Just a thought!

thapav replied:
As discussed over knox, I will mark this comment as resolved

sunghan-chang commented on os/arch/arm/src/common/up_mpuinit.c:
```diff
@@ -50,11 +51,102 @@
  ****************************************************************************/
 
+/****************************************************************************
+ * Name: mpu_regs_initialize
```
Could you check the name?

sunghan-chang commented on os/arch/arm/src/common/up_mpuinit.c:
```diff
@@ -50,11 +51,102 @@
+	offset += MPU_NUM_REGIONS;
+#else
+	offset += MPU_NUM_REGIONS;
+#endif
```
#ifdef and #else are the same. Let's remove the conditional and duplicated line.

sunghan-chang commented on os/arch/arm/src/common/up_mpuinit.c:
```diff
@@ -50,11 +51,102 @@
+	offset += MPU_NUM_REGIONS;
+#else
+	offset += MPU_NUM_REGIONS;
+#endif
```
same as previous

sunghan-chang commented on os/arch/arm/src/common/up_mpuinit.c:
```diff
@@ -50,11 +51,102 @@
+}
+
+/****************************************************************************
+ * Name: mpu_get_region_info
```
check the name

sunghan-chang commented on os/arch/arm/src/common/up_mpuinit.c:
```diff
@@ -50,11 +51,102 @@
+
+uint8_t mpu_get_nregion_info(enum mpu_region_usages_e usages)
+{
+	uint8_t nregion = 0;
```
Let's add `default` instead of zero initialization. Because it can reduce unnecessary operation in the cases of valid usages.

thapav replied:
done

sunghan-chang commented on os/arch/arm/src/common/up_mpuinit.c:
```diff
@@ -50,11 +51,102 @@
+}
+
+/****************************************************************************
+ * Name: mpu_set_region_info
```
name

sunghan-chang commented on os/include/tinyara/mpu.h:
```diff
@@ -101,4 +83,10 @@ bool up_mpu_check_active(uint32_t *mpu_regs);
 
 void up_mpu_disable_region(uint32_t *mpu_regs);
 
+void mpu_region_initialize(FAR struct mpu_usages_s *mpu);
```
could you remove `FAR`?

jeongchanKim reviewed (approved):
LGTM
